src/app/pages/StoryPage/StoryPage.jsx

```diff
@@ -158,9 +159,25 @@ const StoryPage = ({ pageData, mostReadEndpointOverride }) => {
   // ads
   const { enabled: adsEnabled } = useToggle('ads');
   const { enabled: podcastPromoEnabled } = useToggle('podcastPromo');
-  const { isAmp, showAdsBasedOnLocation } = useContext(RequestContext);
+  const { isAmp, showAdsBasedOnLocation, mvtExperiments } =
+    useContext(RequestContext);
+  const { optimizely } = useContext(OptimizelyContext);
   const adcampaign = path(['metadata', 'adCampaignKeyword'], pageData);
 
+  const hasFullStackExperiment =
+    mvtExperiments &&
+    mvtExperiments.some(
+      ({ experimentName, variation, enabled }) =>
+        experimentName === 'full_stack_test' &&
+        variation === 'Control_1' &&
+        enabled,
+    );
+
+  if (hasFullStackExperiment) {
+    activateExperiment(optimizely, 'full_stack_test', 'Control_1');
```

Comment:
Is it possible to figure out the `hasFullStackExperiment` value and do the `activateExperiment(...)` function within the `RequestContext` and then just pass down `hasFullStackExperiment` in `RequestContext` to the page component?

Just thinking it may be clear to abstract this logic out of the page components so that we can just read in a `hasFullStackExperiment` value anywhere we want and know that the value of it will be set.

Reply:
Yeah, I agree the logic should be extracted, basically along the lines of Dario's suggestion. This `hasFullStackExperiment` is a boolean in this instance because there are only 2 variations for this test experiment, this value should actually be the variation. I'm not sure if `RequestContext` is the best place for it, due to thinking about the re-usability side. How about a custom hook, similar to `useOptimizelyVariation`, except in this instance there shouldn't be any re-renders as it's not changing state or mutating the DOM. So like this:

```jsx
const useOptimizelyMvtVariation = id => {
  const { optimizely } = useContext(OptimizelyContext);
  const { mvtExperiments } = useContext(RequestContext);

  const isEnabled =
    mvtExperiments.length !== 0 &&
    mvtExperiments.some(
      ({ experimentName, enabled }) => experimentName === id && enabled,
    );
  const variation = isEnabled && mvtExperiments[id];

  if (variation) activateExperiment(optimizely, id, variation);

  return variation;
};
```

And then in the Story page, or whichever component we want to do the experiment on, we just have to do this
```jsx
const fullStackVariation = useOptimizelyMvtVariation('full_stack_test');

if (fullStackVariation === 'Control_1') {
  const fullStackMessage = `I am the Full Stack Experiment`;
  return <h1>{fullStackMessage}</h1>;
}
```

Reply:
Yea that looks like it could be a good approach. Worth testing the includes rendering issue on http://localhost:7080/mundo/23263889, although there isn't any state updates so it _may_ be ok?

Reply:
Just checked, includes look to be rendering fine, tweaked the hook slightly, will push the changes now. I think `useOptimizelyVariation` uses optimizely's [`useDecision`](https://github.com/bbc/simorgh/blob/72cbdebfef04632d1de257ce5b6c2f81f158913f/src/app/hooks/useOptimizelyVariation/index.jsx#L5-L11), and is set to autoupdate when the datafile resolves, which I think is what caused the re-render bug, not sure what black magic the component variation thing uses, could be something that we could use for the other solution 🤔

Reply:
Yea I'm not entirely sure how the `OptimizelyExperiment` component works. I think the issue is having state changes in the StoryPage component itself, causing the entire component and its children to re-render, so if state changes are happening within the `OptimizelyExperiment` component, thats ok as its isolated to that component only.

Reply:
> Yea I'm not entirely sure how the `OptimizelyExperiment` component works. I think the issue is having state changes in the StoryPage component itself, causing the entire component and its children to re-render, so if state changes are happening within the `OptimizelyExperiment` component, thats ok as its isolated to that component only.

Ooh that makes sense. Maybe we can do that then for the alternative approach? Would need to check to see if it the weird state thing not updating on server side render